File: all_data_structures.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <set>
#include <unordered_set>
#include <map>
#include <unordered_map>
#include <list>
#include <algorithm>
using namespace std;

int main() {
    cout << "Data Structures Fundamentals in C++\n" << endl;

    // 1. Array
    cout << "1. Array\n";
    int arr[] = {1, 2, 3, 4, 5}; // Static array
    cout << "Array elements: ";
    for (int i = 0; i < 5; i++) {
        cout << arr[i] << " ";
    }
    cout << "\n";

    // 2. Vector (Dynamic Array)
    cout << "\n2. Vector (Dynamic Array)\n";
    vector<int> vec = {10, 20, 30};
    vec.push_back(40); // Adding an element
    vec.push_back(50);
    cout << "Vector elements: ";
    for (int val : vec) cout << val << " ";
    cout << "\n";

    // 3. Stack (LIFO)
    cout << "\n3. Stack (LIFO)\n";
    stack<int> stk;
    stk.push(1);
    stk.push(2);
    stk.push(3);
    cout << "Top of stack: " << stk.top() << endl; // Top element
    stk.pop(); // Remove top element
    cout << "Top after pop: " << stk.top() << endl;

    // 4. Queue (FIFO)
    cout << "\n4. Queue (FIFO)\n";
    queue<int> q;
    q.push(5);
    q.push(10);
    q.push(15);
    cout << "Front of queue: " << q.front() << endl;
    q.pop(); // Remove front element
    cout << "Front after pop: " << q.front() << endl;

    // 5. Priority Queue (Heap)
    cout << "\n5. Priority Queue (Heap)\n";
    priority_queue<int> pq; // Max-heap by default
    pq.push(20);
    pq.push(5);
    pq.push(15);
    cout << "Top of max-heap: " << pq.top() << endl;
    pq.pop(); // Remove top
    cout << "Top after pop: " << pq.top() << endl;

    // Min-heap using priority_queue
    priority_queue<int, vector<int>, greater<int>> minHeap;
    minHeap.push(20);
    minHeap.push(5);
    minHeap.push(15);
    cout << "Top of min-heap: " << minHeap.top() << endl;

    // 6. Set (Ordered Collection)
    cout << "\n6. Set (Ordered Collection)\n";
    set<int> s = {10, 20, 10, 30}; // Duplicates are removed
    cout << "Set elements: ";
    for (int val : s) cout << val << " ";
    cout << "\n";

    // 7. Unordered Set (Hash Table)
    cout << "\n7. Unordered Set (Hash Table)\n";
    unordered_set<int> us = {10, 20, 10, 30};
    cout << "Unordered Set elements: ";
    for (int val : us) cout << val << " ";
    cout << "\n";

    // 8. Map (Key-Value Pairs)
    cout << "\n8. Map (Key-Value Pairs)\n";
    map<string, int> mp;
    mp["apple"] = 5;
    mp["banana"] = 3;
    mp["cherry"] = 7;
    cout << "Map elements:\n";
    for (auto [key, value] : mp) {
        cout << key << ": " << value << endl;
    }

    // 9. Unordered Map (Hash Table for Key-Value Pairs)
    cout << "\n9. Unordered Map (Hash Table)\n";
    unordered_map<string, int> ump;
    ump["car"] = 10;
    ump["bike"] = 20;
    ump["bus"] = 30;
    cout << "Unordered Map elements:\n";
    for (auto [key, value] : ump) {
        cout << key << ": " << value << endl;
    }

    // 10. Linked List
    cout << "\n10. Linked List\n";
    list<int> linkedList = {1, 2, 3, 4, 5};
    linkedList.push_front(0); // Add to front
    linkedList.push_back(6); // Add to back
    cout << "Linked List elements: ";
    for (int val : linkedList) cout << val << " ";
    cout << "\n";

    // 11. Deque (Double-Ended Queue)
    cout << "\n11. Deque (Double-Ended Queue)\n";
    deque<int> dq = {1, 2, 3};
    dq.push_front(0); // Add to front
    dq.push_back(4); // Add to back
    cout << "Deque elements: ";
    for (int val : dq) cout << val << " ";
    cout << "\n";

    // 12. Binary Search (Algorithm)
    cout << "\n12. Binary Search\n";
    vector<int> searchVec = {1, 3, 5, 7, 9};
    int key = 5;
    bool found = binary_search(searchVec.begin(), searchVec.end(), key);
    cout << "Is " << key << " in the vector? " << (found ? "Yes" : "No") << endl;

    // 13. Sorting (Algorithm)
    cout << "\n13. Sorting\n";
    vector<int> sortVec = {5, 3, 8, 1, 2};
    sort(sortVec.begin(), sortVec.end());
    cout << "Sorted vector: ";
    for (int val : sortVec) cout << val << " ";
    cout << "\n";

    return 0;
}

File: alldatastructures.py
# Major data structures demo
# Written by Tristan Onek

# 1. Lists:
print("1. List (Dynamic array)")
my_list = [1, 2, 3, 4]
# Adding an element to the list
my_list.append(5)
# Lists are mutable and maintain order
print("After append:", my_list)  # Output: [1, 2, 3, 4, 5]

# 2. Tuples:
print("\n2. Tuple (Immutable array)")
my_tuple = (1, 2, 3, 4)
# Accessing an element
print("Element at index 2:", my_tuple[2])  # Output: 3
# Tuples are immutable; you can't modify them

# 3. Sets:
print("\n3. Set (Unordered collection of unique elements)")
my_set = {1, 2, 2, 3, 4}
# Adding an element
my_set.add(5)
print("After adding 5:", my_set)  # Output: {1, 2, 3, 4, 5}

# 4. Dictionary (KVPs):
print("\n4. Dictionary (Key value pairs)")
my_dict = {"a": 1, "b": 2, "c": 3}
# Adding a new key-value pair
my_dict["d"] = 4
print("...After adding ('d': 4):", my_dict)  # Output: {'a': 1, 'b': 2, 'c': 3, 'd': 4}

# 5. Stack:
print("\n5. Stack (e.g. LIFO)")
stack = []
# Pushing elements onto the stack...
stack.append(1)
stack.append(2)
stack.append(3)
print("Stack after pushing:", stack)  # Output: [1, 2, 3]
# Popping an element
print("Popped element:", stack.pop())  # Output: 3
print("Stack after popping:", stack)  # Output: [1, 2]

# 6. Queue:
print("\n6. Queue (FIFO)")
from collections import deque
queue = deque()
# Queue elements
queue.append(1)
queue.append(2)
queue.append(3)
print("Queue after enqueues:", queue)  # Output: deque([1, 2, 3])
# Dequeue an element
print("Dequeued element:", queue.popleft())  # Output: 1
print("After dequeue:", queue)  # Output: deque([2, 3])

# 7. Priority Queue:
print("\n7. Priority Queue (heapq)")
import heapq
priority_queue = []
# Adding elements
heapq.heappush(priority_queue, 3)
heapq.heappush(priority_queue, 1)
heapq.heappush(priority_queue, 2)
print("Priority queue after pushes:", priority_queue)  # Output: [1, 3, 2]
# Popping the smallest element
print("Popped smallest element:", heapq.heappop(priority_queue))  # Output: 1
print("Priority queue after pop:", priority_queue)  # Output: [2, 3]

# 8. Linked List:
print("\n8. Linked List")
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

linked_list = LinkedList()
linked_list.append(1)
linked_list.append(2)
linked_list.append(3)
print("Linked list:")
linked_list.print_list()  # Output: 1 -> 2 -> 3 -> None

# 9. Graph Adjacency list demo:
print("\n9. Graph (Adjacency list)")
graph = {
    0: [1, 2],
    1: [0, 3],
    2: [0, 3],
    3: [1, 2]
}

print("Graph adjacency list:", graph)
# Access neighbors of a node
print("Neighbors of node 2:", graph[2])  # Output: [0, 3]

# 10. Binary Tree:
print("\n10. Binary Tree")
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self, root_value):
        self.root = TreeNode(root_value)

    def insert_left(self, current_node, value):
        current_node.left = TreeNode(value)

    def insert_right(self, current_node, value):
        current_node.right = TreeNode(value)

    def inorder_traversal(self, node):
        if node:
            self.inorder_traversal(node.left)
            print(node.value, end=" ")
            self.inorder_traversal(node.right)

tree = BinaryTree(1)
tree.insert_left(tree.root, 2)
tree.insert_right(tree.root, 3)
print("In-order traversal of tree:")
tree.inorder_traversal(tree.root)  # Output: 2 1 3
print()
